exporter/prometheusremotewriteexporter/exporter_test.go

```diff
@@ -997,69 +997,82 @@ func TestWALOnExporterRoundTrip(t *testing.T) {
 	assert.Equal(t, gotFromWAL, gotFromUpload)
 }
 
-func TestRetryOn5xx(t *testing.T) {
-	// Create a mock HTTP server with a counter to simulate a 5xx error on the first attempt and a 2xx success on the second attempt
-	attempts := 0
-	mockServer := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		if attempts < 4 {
-			attempts++
-			http.Error(w, "Internal Server Error", http.StatusInternalServerError)
-		} else {
-			w.WriteHeader(http.StatusOK)
-		}
-	}))
-	defer mockServer.Close()
-
-	endpointURL, err := url.Parse(mockServer.URL)
-	require.NoError(t, err)
-
-	// Create the prwExporter
-	exporter := &prwExporter{
-		endpointURL: endpointURL,
-		client:      http.DefaultClient,
-		retrySettings: exporterhelper.RetrySettings{
-			Enabled: true,
-		},
-	}
-
-	ctx := context.Background()
-
-	// Execute the write request and verify that the exporter returns a non-permanent error on the first attempt.
-	err = exporter.execute(ctx, &prompb.WriteRequest{})
-	assert.NoError(t, err)
-	assert.Equal(t, 4, attempts)
+func canceledContext() context.Context {
+	ctx, cancel := context.WithCancel(context.Background())
+	cancel()
+	return ctx
 }
 
-func TestNoRetryOn4xx(t *testing.T) {
-	// Create a mock HTTP server with a counter to simulate a 4xx error
-	attempts := 0
-	mockServer := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		if attempts < 1 {
-			attempts++
-			http.Error(w, "Bad Request", http.StatusBadRequest)
-		} else {
-			w.WriteHeader(http.StatusOK)
-		}
-	}))
-	defer mockServer.Close()
+func TestRetries(t *testing.T) {
 
-	endpointURL, err := url.Parse(mockServer.URL)
-	require.NoError(t, err)
-
-	// Create the prwExporter
-	exporter := &prwExporter{
-		endpointURL: endpointURL,
-		client:      http.DefaultClient,
-		retrySettings: exporterhelper.RetrySettings{
-			Enabled: true,
+	tts := []struct {
+		name             string
+		serverErrorCount int // number of times server should return error
+		expectedAttempts int
+		httpStatus       int
+		shouldError      bool
```

Comment:
You can use the `assert.Error` interface here rather than a `shouldError` flag. Then you can just use the assertion function you expect. 

Based on your test logic you will most likely have to add a `isPermanent` field that is also an `assert.Error` interface. But that should be okay.

Reply:
TIL.
thanks!